src/azul/deployment.py

```diff
@@ -277,5 +285,158 @@ def _sanitize_tf(tf_config: JSON) -> JSON:
+                   tags or to override default tags.
+    """
+    component_suffix, _ = config.unqualified_resource_name_or_none(resource_name)
+    if component_suffix is None and 'component' not in kwargs:
+        raise RequirementError(f"Component tag could not be inferred from {resource_name=}."
+                               " It should be specified explicitly.")
+    else:
+        return {
+            'project': 'dcp',
+            'service': config.resource_prefix,
+            'deployment': config.deployment_stage,
+            'owner': config.owner,
+            'name': resource_name,
+            'component': f'{config.resource_prefix}-{component_suffix}',
+            **kwargs
+        }
+
+
+def add_tags(resource_type: str, resource_config: JSON) -> JSON:
+    """
+    Given a Terraform resource, add resource tags based on a name
+    attribute present in the `resource_config`. Existing tags are retained.
+    Resources that do not support tagging are returned with their configuration
+    unchanged.
+
+    :param str resource_type: the Terraform resource type associated with the
+                              resource, e.g., `aws_acm_certificate`
+    :param JSON resource_config: Terraform resource configuration, e.g., `foo`
+                                 in `{"resource_type": {"resource_name": foo}}`
+    """
+    name_key_by_resource = {  # resource_config key containing resource name, or None to skip
+        'aws_acm_certificate': 'domain_name',
+        'aws_acm_certificate_validation': None,
+        'aws_api_gateway_base_path_mapping': None,
+        'aws_api_gateway_deployment': None,
+        'aws_api_gateway_domain_name': 'domain_name',
+        'aws_cloudwatch_event_target': None,
+        'aws_iam_instance_profile': None,
+        'aws_iam_policy': None,
+        'aws_iam_role_policy': None,
+        'aws_iam_role_policy_attachment': None,
+        'aws_key_pair': 'key_name',
+        'aws_lambda_event_source_mapping': None,
+        'aws_lambda_function': 'function_name',
+        'aws_lambda_layer_version': None,
+        'aws_lambda_permission': None,
+        'aws_lb_listener': None,
+        'aws_lb_target_group_attachment': None,
+        'aws_route_table_association': None,
+        'aws_route53_record': None,
+        'aws_s3_bucket': 'bucket',
+        'aws_volume_attachment': None,  # Resource supports tagging but Terraform rejects
+        'google_service_account': None,
+        'null_resource': None,
+        # These resources support tagging but require a manually-specified name
+        'aws_eip': None,
+        'aws_instance': None,
+        'aws_internet_gateway': None,
+        'aws_nat_gateway': None,
+        'aws_network_interface': None,
+        'aws_route': None,
+        'aws_route_table': None,
+        'aws_subnet': None,
+        'aws_vpc': None
+    }
+    name_key = name_key_by_resource.get(resource_type, "name")
```

Comment:
This is more or less the current behavior, though looking again I can make it more obvious what code needs to be modified and how.

The slight deviation from your description is that `add_tags` will try and infer tagging information and silently add it if possible. If `add_tags` can't do that (i.e., if the resource doesn't have a `name` key that parses nicely) then it will throw an error when Terraform configuration is being generated/validated. Is this the correct behavior?

Reply:
Yes.